Modern Testing

Rot127 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [ ] I've documented or updated the documentation of every API function and struct this PR changes.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

Supersedes https://github.com/capstone-engine/capstone/pull/2456

## Modernize testing.

The new unit testing is build around `yaml` files. They can be consumed by the C test engine (`cstest`) or any bindings easily.

### Statistics

The new `cstest` logs technically two statistics. One by `cmocka` and one by `cstest`.
The reason is:
1. We might want to group tests in the future (by arch or file for example) so we need a global statistics.
2. Errors of incorrect yaml parsing are not tracked by cmocka.
3. Also, `cmocka` doesn't provide the detailed statistics to us. So we cannot print a nice summary at the end.

### Changes

- New `cstest` C-implementation
- New `py_cstest` Python-implementation
- Moves almost all tests to `yaml` files.
- Old tests which can't be moved (because they are effectively integration tests) are kept.
- Add numerous test jobs in the workflow
- Adds several CMake test targets
- Add LoongArch Python bindings
- Several bug fixes discovered during testing.

Current test coverage of Auto-Sync `<ARCH>Mapping.c` files by running `cstest test/`:

| file                                                          | coverage | covered     |
|---------------------------------------------------------------|----------|-------------|
| arch/LoongArch/LoongArchMapping.c                             | 67.29%   | 144 / 214   |
| arch/PowerPC/PPCMapping.c                                     | 72.78%   | 254 / 349  
| arch/AArch64/AArch64Mapping.c                                 | 60.47%   | 918 / 1518  |
| arch/TriCore/TriCoreMapping.c                                 | 45.00%   | 45 / 100    |
| arch/Alpha/AlphaMapping.c                                     | 86.08%   | 68 / 79     |
| arch/ARM/ARMMapping.c                                         | 61.68%   | 792 / 1284  |
| arch/HPPA/HPPAMapping.c                                       | 29.85%   | 20 / 67     |

**After merge**

- [ ] Add windows build to CI

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Or what test cases you added. Disassembly tests should be added in suite/cs_test/issues.cs -->

Added and all green

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

closes https://github.com/capstone-engine/capstone/issues/1984
closes https://github.com/capstone-engine/capstone/issues/2405
closes https://github.com/capstone-engine/capstone/issues/2383
closes https://github.com/capstone-engine/capstone/issues/2383

Rot127 committed:
Replace old cstest with new implementation.

Rot127 committed:
Remove and partially move legacy tests.

- Not removed are the legacy integration tests.
- Detail and asm tests are now in the yaml test files.

Rot127 committed:
Rewrite MCUpdater

The MCUpdater is completely rewritten to generate yaml test files
from the LLVM MC tests.

Also the compatibility header got a test case in CMake.

Rot127 committed:
Add new tests files.

Rot127 committed:
Remove unused files.

Rot127 committed:
Fix several bugs and clang-tidy warnings detected by the new tests.

Rot127 committed:
Add py_cstest and LoongArch Python bindings.

Rot127 committed:
Update docs and Github workflows.

Rot127 commented:
I am aware that the Auto-Sync test fails. I have no idea why (it works fine locally) but fix it later. Just wanted this ready for review.

The clang-tidy issue is a false positive. But I cannot for the sake of it create a minimal working example to report the bug.

I am not sure if I find the time to try it again. But either we force merge or silence it.

XVilka commented on CMakeLists.txt:
```diff
@@ -766,27 +751,12 @@ if(BUILD_SHARED_LIBS)
+# Fuzzer if this is moved to it's own CMakeLists.txt (as it should be)
+# the OSS fuzzer build fails. And must be fixed.
+# Simply because it builds the fuzzer there again with hard-coded paths.
+# See: https://github.com/google/oss-fuzz/blob/master/projects/capstone/build.sh
```
Please open an issue and refer to it here in the code, so that it is easier to track and not forget

XVilka commented on COMPILE_MAKE.TXT:
```diff
@@ -1,3 +1,11 @@
+
 This documentation explains how to compile, install & run Capstone on MacOSX,
 Linux, *BSD & Solaris. We also show steps to cross-compile for Microsoft Windows.
 
```
Maybe also makes sense to mark COMPILE_MSVC as deprecated too? Since CMake can handle Windows just fine as well, and latest versions of VS and VSCode work with CMake quite well.

XVilka replied:
Maybe even XCode build as well.

Rot127 replied:
Yes, but I would do it in another PR. I also did not edited or removed the MSVC project files. Because it gets too bloated otherwise.
But opened an issue about it: https://github.com/capstone-engine/capstone/issues/2457

XVilka commented on README.md:
```diff
@@ -49,7 +49,7 @@ Further information is available at https://www.capstone-engine.org
 -------
 
-See COMPILE.TXT file for how to compile and install Capstone.
+See COMPILE_CMAKE.TXT file for how to compile and install Capstone.
```
Could be a link maybe? Since it's Markdown anyway.

XVilka commented on arch/AArch64/AArch64InstPrinter.c:
```diff
@@ -2070,7 +2076,7 @@ void printAdrpLabel(MCInst *MI, uint64_t Address, unsigned OpNum, SStream *O)
 	}
 
-	SStream_concat0(O, "<llvm-expr>");
+	printUInt64Bang(O, MCOperand_getImm(Op));
```
Interesting catch. I wonder how it worked before.

Rot127 replied:
This was new and still works either way. I added it there a few weeks ago to show the user that the operand gets processed somehow by LLVM.

But in practice it is barely reached and it turns out LLVM just prints 0 in this case. So here it is reverted back to this behavior.

XVilka commented on arch/AArch64/AArch64Mapping.c:
```diff
@@ -2500,7 +2525,7 @@ void AArch64_set_detail_op_imm(MCInst *MI, unsigned OpNum,
 			AArch64_set_detail_op_sme(MI, OpNum,
 						  AARCH64_SME_MATRIX_SLICE_OFF,
-						  AARCH64LAYOUT_INVALID, 1);
+						  AARCH64LAYOUT_INVALID, (int64_t) 1);
```
Alternatively you could use something like `1LL`

Rot127 replied:
Is `LL` always equivalent to a 64bit signed integer?

XVilka replied:
Yes, it's a shorthand for `long long`

XVilka commented on arch/AArch64/AArch64Mapping.c:
```diff
@@ -2534,8 +2559,8 @@ void AArch64_set_detail_op_imm_range(MCInst *MI, unsigned OpNum,
-						  AARCH64LAYOUT_INVALID, FirstImm,
-						  Offset);
+						  AARCH64LAYOUT_INVALID, (int) FirstImm,
+						  (int) Offset);
```
Why here it cast to `int` while in the previous change the last argument was of the `int64_t` type? Also using `int` for offset looks suspicious

Rot127 replied:
I use `int8_t` in the target value, because I need to have an invalid value like -1. But yeah. I should go with `uint16_t` and invalid is `UINT16_MAX`

XVilka commented on bindings/python/BUILDING.txt:
````diff
@@ -4,17 +4,12 @@
-		$ sudo make install
-
-   To control the install destination, set the DESTDIR environment variable.
+```
````
If you change the syntax, you need also rename the file into `BUILDING.md`

XVilka commented on bindings/python/setup.py:
```diff
@@ -72,10 +72,12 @@
     LIBRARY_FILE = "libcapstone.so"
     STATIC_LIBRARY_FILE = 'libcapstone.a'
```
This code could probably be simplified to use the name only once. But it's probably better for another PR

Rot127 replied:
Yes. The whole thing is tracked in https://github.com/capstone-engine/capstone/issues/2445

XVilka commented on bindings/python/setup.py:
```diff
@@ -134,23 +133,22 @@ def build_libraries():
+    print("Build Directory: {}\n".format(os.getcwd()))
+    # Only build capstone.dll / libcapstone.dylib
+    if SYSTEM == "win32":
+        os.system('cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON -DCAPSTONE_BUILD_LEGACY_TESTS=OFF -DCAPSTONE_BUILD_CSTOOL=OFF -G "NMake Makefiles" ..')
```
Some of these could be deduplicated too, I think. But also fine for another PR

Rot127 replied:
As above: https://github.com/capstone-engine/capstone/issues/2445

XVilka commented on suite/auto-sync/src/autosync/Targets.py:
```diff
@@ -0,0 +1,4 @@
+# Copyright © 2024 Rot127 <unisono@quyllur.org>
+# SPDX-License-Identifier: BSD-3
+
+TARGETS_LLVM_NAMING = ["ARM", "PowerPC", "Alpha", "AArch64", "LoongArch"]
```
Missing Tricore?

Rot127 replied:
Actually yes. But we also don't have it in `llvm-capstone`. @imbillow How did we proceed back then? Should we add it?

XVilka commented on include/capstone/arm.h:
```diff
@@ -903,7 +903,8 @@ typedef struct cs_arm {
 	bool update_flags;	///< does this insn update flags?
 	bool post_index;	///< only set if writeback is 'True', if 'False' pre-index, otherwise post.
-	int /* arm_mem_bo_opt */ mem_barrier;	///< Option for some memory barrier instructions
+	arm_mem_bo_opt
```
Looks like broken indent?

XVilka commented on utils.c:
```diff
@@ -125,4 +125,46 @@ void append_to_str_lower(char *str, size_t str_size, const char *src) {
+/// @param str The string to append to.
+/// @param str_buf_size Size of buffer @p str.
+/// @param src The string to append.
+void append_to_str(char *str, size_t str_buf_size, const char *src) {
```
Check for NULL first.

XVilka commented on utils.c:
```diff
@@ -125,4 +125,46 @@ void append_to_str_lower(char *str, size_t str_size, const char *src) {
+
+/// Returns the given byte sequence @bytes as a string of the
+/// form: 0xXX,0xXX...
+char *byte_seq_to_str(uint8_t *bytes, size_t len)
```
Check if bytes is NULL as well

XVilka commented on tests/README.md:
```diff
@@ -0,0 +1,158 @@
+Legacy tests which only printed to `stdout`. In practice they only test if the code segfaults.
+Checking the produced output was not implemented.
+
+_YAML test files_
```
Move legacy to the end, after main YAML tests

Rot127 committed:
Fix documentation related review requests.

Rot127 committed:
Add missing NULL checks

Rot127 committed:
Fix intend.

Rot127 committed:
Make slice operand members uints and fix funny casting of them.

Rot127 committed:
Remove any possibility to run tests in parallel.

Rot127 committed:
Install and document FileCheck usage

Rot127 committed:
Delete unused file

kabeor commented on .github/workflows/CITest.yml:
```diff
@@ -34,41 +34,53 @@ jobs:
+            }
+          - {
+              name: 'ubuntu-latest x64 cmake',
+              os: ubuntu-latest,
```
ubuntu-latest is Ubuntu 22.04 according to https://github.com/actions/runner-images, which is duplicate with `ubuntu-22.04 x64 cmake` following. How about we only run cmake on ubuntu-24.04.

Rot127 committed:
Update saved_patches.json

Rot127 committed:
Remove Ubuntu latest job as duplicate for Ubuntu 22.04

Rot127 committed:
Update Differ saved_choices check

Rot127 committed:
Set tree-sitter version to 0.22.3 because they removed deprecated functions we use.

Rot127 commented:
@kabeor I fixed the Auto-Sync workflow. Please take another look.

Rot127 committed:
Allow to add an mandatory option to the generated MC tests.

Rot127 committed:
Fix: Skip tests which test for symbols.

Rot127 committed:
Renaming py_cstest -> cstest_py

Rot127 committed:
Don't generate tests from assembly tests.

We cannot toggle between alias or non aias assembly currently.
So valid tests fail because the non alias asm text is compared.

Rot127 committed:
Suppress false positive uninitialized warning.

kabeor reviewed (approved):
Cool!
